### ISSUE TITLE
React ref union types cause false errors

### ISSUE THREAD

**Issue by just-boris** [microsoft/TypeScript#36659]
## Search Terms

react ref union type

## Suggestion

Currently React refs do not work when you applying it to different elements depending on some condition. I'd like to not have Typescript errors for this case

## Use Cases

Sometimes we need to conditionally render different elements and attach refs to them:

```jsx
function Button({ href }: { href: string }) {
    const ref = React.useRef<HTMLAnchorElement | HTMLButtonElement>()
    if (href) {
        return <a ref={ref}></a>
//               Type 'RefObject<HTMLAnchorElement | HTMLButtonElement>' is not assignable
//               to type 'RefObject<HTMLAnchorElement>'.
    }
    return <button ref={ref}></button>

}
```

This code is valid Javascript, but reports an error in Typescript.

[Playground example](http://www.typescriptlang.org/play/index.html?strictBindCallApply=false&jsx=2#code/PTAEBsEsBdvBTUBnSBbAruAhtSB7AO1DwDNQAleLAY2lGgE8AHeJAKABN5rsAnRAllSsmNRJRp0A3m1BzQ8AB5M8vOpALR4vEmIrwSAeQBGAK27QAPABUAfKBnynoflg6FwDUNXS9+mgC5Qa1AAH1ACTHAAbll5AF84uSUVNVASdAJafCJqVy0AUQRhTRsFRS0CDiRQAGt4BlJQACkAZQANADoASU1eDRRqIvgS6CRbAApGFiDrABpQJl48JiQgtq7e6H6CQeHRpABKIIA3PEgOWKcU1ToMrNxCUHQkeEoSG0nj-SMzC0+2Ik2IJhEhRNREBsHElQBotDo9AAJawAWQAMgBBWD9YzoLRIT7Q5zyfgkIISWidd4mcy0AFORLXZS3WGabS6CGgLY7PbFeCaGqOYmgLBBZHorHbSC4-GWcWYrIAC1U+350FsMKcMughDFqMx2OleNYcv1ACE8TqCKrNBqGYC2I6QPRFZAagB3VS1JALYzcLAvRDQRXaRBukVEJSSUCoHDURWgP3Qd3wfkuAz0ZiILBVBR8zRse7ZJ6tNBMBAW2CECaHInyaiEJB0UmgAC8+kknUD71N6MrVpt6prMP40F8REs2qepNbUlJ8VslmAU4IGodbGdwfD7lYETwdE9vFqvv9gZdQazMawXhOWF4DELmWLRH71akoEVLfiQXfn4MQSbHYAHNQHiWshTkBtdmbDN2wpaAu1eHt5QxJUVXzOhwnlV9rQwr4YUgMgJj-EhwM1El4DHXgJywdMSFnedF2ALA7QSEdKPHUBJ0tacDAYgwFyXFc13iIA)

I think, this is also related to the general covariance issue: https://github.com/microsoft/TypeScript/issues/10717#issuecomment-461977080
But I am wondering if there could be a more simple solution to this particular case

**Comment by nmain100** [comment]
https://github.com/Microsoft/TypeScript/issues/21759 could solve this, as when passing a ref object to React.createElement, only the write side is consumed.

**Comment by just-boris** [comment]
Solved my issue via `React.useImperativeHandle` hook:

```ts
function Button({ href }: { href: string }, ref: React.Ref<{focus(): void}>) {
    const ref = React.useRef<any>()
    React.useImperativeHandle(ref, () => ({focus: () => ref.current && ref.current.focus}))
    if (href) {
        return <a ref={ref}></a>
    }
    return <button ref={ref}></button>

}
```
